src/test/sql/sql_pipeline_test.cpp

```diff
@@ -0,0 +1,470 @@
+  EXPECT_TRUE(_find_in_lqp(lqp_roots.at(0), _contains_cross));
+}
+
+TEST_F(SQLPipelineTest, GetUnoptimizedLQPInvalid) {
+  SQLPipeline sql_pipeline{_invalid_sql};
+  EXPECT_THROW(sql_pipeline.get_unoptimized_logical_plans(), std::exception);
+}
+
+TEST_F(SQLPipelineTest, GetUnoptimizedLQPValidated) {
+  SQLPipeline sql_pipeline{_select_query_a};
+
+  const auto& lqp_roots = sql_pipeline.get_unoptimized_logical_plans();
+
+  EXPECT_EQ(lqp_roots.size(), 1u);
+  EXPECT_TRUE(lqp_roots.at(0)->subtree_is_validated());
+}
+
+TEST_F(SQLPipelineTest, GetUnoptimizedLQPNotValidated) {
+  SQLPipeline sql_pipeline{_select_query_a, false};
+
+  const auto& lqp_roots = sql_pipeline.get_unoptimized_logical_plans();
+
+  EXPECT_EQ(lqp_roots.size(), 1u);
+  EXPECT_FALSE(lqp_roots.at(0)->subtree_is_validated());
+}
+
+TEST_F(SQLPipelineTest, GetOptimizedLQP) {
+  SQLPipeline sql_pipeline{_select_query_a};
+
+  const auto& lqp_roots = sql_pipeline.get_optimized_logical_plans();
+
+  EXPECT_EQ(lqp_roots.size(), 1u);
+  EXPECT_EQ(lqp_roots.at(0)->output_column_names(), _column_names);
+}
+
+TEST_F(SQLPipelineTest, GetOptimizedLQPTwice) {
+  SQLPipeline sql_pipeline{_select_query_a};
+
+  sql_pipeline.get_optimized_logical_plans();
+  const auto& lqp_roots = sql_pipeline.get_optimized_logical_plans();
+
+  EXPECT_EQ(lqp_roots.size(), 1u);
+  EXPECT_EQ(lqp_roots.at(0)->output_column_names(), _column_names);
+}
+
+TEST_F(SQLPipelineTest, GetOptimizedLQPWithJoinFilter) {
+  SQLPipeline sql_pipeline{_join_query};
+
+  const auto& lqp_roots = sql_pipeline.get_optimized_logical_plans();
+
+  EXPECT_EQ(lqp_roots.size(), 1u);
+  EXPECT_FALSE(_find_in_lqp(lqp_roots.at(0), _contains_cross));
+  EXPECT_EQ(lqp_roots.at(0)->output_column_names(), _join_column_names);
+}
+
+TEST_F(SQLPipelineTest, GetOptimizedLQPInvalid) {
+  SQLPipeline sql_pipeline{_invalid_sql};
+  EXPECT_THROW(sql_pipeline.get_optimized_logical_plans(), std::exception);
+}
+
+TEST_F(SQLPipelineTest, GetOptimizedLQPValidated) {
+  SQLPipeline sql_pipeline{_select_query_a};
+
+  const auto& lqp_roots = sql_pipeline.get_optimized_logical_plans();
+
+  EXPECT_EQ(lqp_roots.size(), 1u);
+  EXPECT_TRUE(lqp_roots.at(0)->subtree_is_validated());
+}
+
+TEST_F(SQLPipelineTest, GetOptimizedLQPNotValidated) {
+  SQLPipeline sql_pipeline{_select_query_a, false};
+
+  const auto& lqp_roots = sql_pipeline.get_optimized_logical_plans();
+
+  EXPECT_EQ(lqp_roots.size(), 1u);
+  EXPECT_FALSE(lqp_roots.at(0)->subtree_is_validated());
+}
+
+TEST_F(SQLPipelineTest, GetOptimizedLQPDoesNotInfluenceUnoptimizedLQP) {
+  SQLPipeline sql_pipeline{_join_query};
+
+  const auto& unoptimized_lqp_roots = sql_pipeline.get_unoptimized_logical_plans();
+
+  // The optimizer works on the original LQP nodes which could be modified during optimization.
+  // Copy the structure to check that it is equal after optimizing.
+  std::vector<std::shared_ptr<AbstractLQPNode>> unoptimized_copies;
+  for (const auto& root : unoptimized_lqp_roots) {
```

Comment:
Further up you're asserting that there is just one element, here you're iterating over this one element list. Apart from this inconsistency, I don't like loops in tests (l 277) because on failure they never tell you on **which** index the EXPECT_* failed.